addons/binding/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/ICloudConnection.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.icloud.internal;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Base64;
+import java.util.Properties;
+
+import org.eclipse.smarthome.io.net.http.HttpUtil;
+import org.openhab.binding.icloud.internal.json.request.ICloudAccountDataRequest;
+import org.openhab.binding.icloud.internal.json.request.ICloudFindMyDeviceRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * Handles communication with the Apple server. Provides methods to
+ * get device information and to find a device.
+ *
+ * @author Patrik Gfeller - Initial Contribution
+ *
+ */
+public class ICloudConnection {
+    private final byte[] authorization;
+    private final Gson gson = new GsonBuilder().create();
+    private Properties httpHeader;
+    private final String iCloudAPIPingDeviceCommand = "/playSound";
+    private final String iCloudAPIRequestDataCommand = "/initClient";
+    private final String iCloudApiURL = "https://fmipmobile.icloud.com/fmipservice/device/";
+    private final String iCloudDataRequest = gson.toJson(ICloudAccountDataRequest.defaultInstance());
+    private URL iCloudDataRequestURL;
+    private URL iCloudFindMyDeviceURL;
+    private final String iCloudURL = "https://www.icloud.com";
+    private final Logger logger = LoggerFactory.getLogger(ICloudConnection.class);
+    private final int socketTimeout = 2500;
+
+    public ICloudConnection(String appleId, String password) throws MalformedURLException {
+        authorization = Base64.getEncoder().encode((appleId + ":" + password).getBytes());
```

Comment:
I think its purer and less fragile if you pass this authorization or the details to `createHttpHeader` (and maybe make that one static) as well as `getBasicAuthorization`. Because by calling those methods from the constructor they operate on a not fully constructed objected. And given `this.authorization` is only used through constructor -> createHttpHeader -> getBasicAuthorization we don't need to store it but can just pass it in thus don't risk future bugs of calling non-initialized fields.

Reply:
... done.